### PR TITLE
Feature alb subnets

### PR DESCRIPTION
**Pull request by jerryhe1999**
### Issue

#3082 

### Description

Added a new feature gate named "ALBSingleSubnet" with default value false, once it set to true, the user who get whitelisted by AWS ELB team for using only one subnet for their application load balancer could be processed as expected.

Manual Test Items:

For account which doesn't get whitelisted, creating the ingress resource with only one subnet attached. 
Comes up with LBC error message while ALBSIngleSubnet is set to false.
Comes up with ELB error message while ALBSIngleSubnet is set to true.

For account which get whitelisted, creating the ingress resource with only one subnet attached. 
Comes up with LBC error message while ALBSIngleSubnet is set to false.
ALB is successfully provisioning while ALBSIngleSubnet is set to true.

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [X] Manually tested
- [X] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/networking/subnet_resolver.go [hunk 1]
```diff
@@ -48,6 +48,8 @@ type SubnetsResolveOptions struct {
 	AvailableIPAddressCount int64
 	// whether to check the cluster tag
 	SubnetsClusterTagCheck bool
+	// Disable subnet minimal count restriction
```
Comment (oliviassss) [inline pkg/networking/subnet_resolver.go:51]: Maybe reword: whether to allow using only 1 subnet for provisioning ALB, default to `false`

File: pkg/networking/subnet_resolver.go [hunk 2]
```diff
@@ -95,6 +97,13 @@ func WithSubnetsClusterTagCheck(SubnetsClusterTagCheck bool) SubnetsResolveOptio
 	}
 }
 
+// WithALBSingleSubnet generate an option that foncigure ALBSingleSubnet
```
Comment (oliviassss) [inline pkg/networking/subnet_resolver.go:100]: nit: foncigure or configure?

File: pkg/networking/subnet_resolver.go [hunk 3]
```diff
@@ -364,6 +373,9 @@ func (r *defaultSubnetsResolver) validateSubnetsMinimalCount(subnets []*ec2sdk.S
 // computeSubnetsMinimalCount returns the minimal count requirement for subnets.
 func (r *defaultSubnetsResolver) computeSubnetsMinimalCount(subnetLocale subnetLocaleType, resolveOpts SubnetsResolveOptions) int {
 	minimalCount := 1
+	if resolveOpts.ALBSingleSubnet {
```
Comment (oliviassss) [inline pkg/networking/subnet_resolver.go:376]: Maybe add a check that the LBType is `elbv2model.LoadBalancerTypeApplication`

**Comment by oliviassss** [comment]
Thanks for the contribution. Please also add:
1. doc updates regarding the new feature gate flag: https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/docs/deploy/configurations.md#feature-gates
2. doc updates re subnet discovery: https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/docs/deploy/subnet_discovery.md#subnet-auto-discovery
3. Please add the manual tests in the PR descriptions.

File: pkg/networking/subnet_resolver.go (hunk 3 above)
Comment (jerryhe1999) [reply pkg/networking/subnet_resolver.go:376]: I am not sure whether is necessary to add the check of elbv2model.LoadBalancerTypeApplication here since if the type is NLB the minimalCount will be returned with 1 anyway, but it might make this code change more readable?

File: pkg/networking/subnet_resolver.go [hunk 4]
```diff
@@ -364,6 +373,9 @@ func (r *defaultSubnetsResolver) validateSubnetsMinimalCount(subnets []*ec2sdk.S
 // computeSubnetsMinimalCount returns the minimal count requirement for subnets.
 func (r *defaultSubnetsResolver) computeSubnetsMinimalCount(subnetLocale subnetLocaleType, resolveOpts SubnetsResolveOptions) int {
 	minimalCount := 1
+	if resolveOpts.ALBSingleSubnet && resolveOpts.LBType == elbv2model.LoadBalancerTypeApplication {
+		return minimalCount
+	}
 	if resolveOpts.LBType == elbv2model.LoadBalancerTypeApplication && subnetLocale == subnetLocaleTypeAvailabilityZone {
```
Comment (johngmyers) [inline pkg/networking/subnet_resolver.go:379]: Would be much more straightforward as:

```suggestion
	if resolveOpts.LBType == elbv2model.LoadBalancerTypeApplication && subnetLocale == subnetLocaleTypeAvailabilityZone && !resolveOpts.ALBSingleSubnet {
```

**Comment by johngmyers** [comment]
Is there any particular reason this needs to be feature-flagged?

**Comment by johngmyers** [comment]
To answer my own question, I think this is so that LBC can give better and earlier error messages in the common case where the ALB service requires a minimum of 2.

**Comment by oliviassss** [comment]
@johngmyers, it's an opt-in feature from ELB side. They are working on lifting the minimal subnet constraint for ALB. AFAIK, in some region like KIX they only require 1 subnet for ALB, but most regions still require 2 subnets as minimal unless the account is allowlisted.

File: pkg/networking/subnet_resolver.go (hunk 4 above)
Comment (jerryhe1999) [reply pkg/networking/subnet_resolver.go]: Thank you for reviewing @johngmyers, the change has been made and submitted.
